ui/ozone/platform/x11/x11_display_manager.cc

```diff
@@ -0,0 +1,335 @@
+    gfx_display.SetScaleAndBounds(device_scale_factor, bounds_in_pixels);
+  }
+
+  return std::vector<display::Display>(1, gfx_display);
+}
+
+}  // namespace
+
+////////////////////////////////////////////////////////////////////////////////
+// X11DisplayManager, public:
+
+X11DisplayManager::X11DisplayManager()
+    : xdisplay_(gfx::GetXDisplay()),
+      x_root_window_(DefaultRootWindow(xdisplay_)),
+      xrandr_version_(0),
+      xrandr_event_base_(0),
+      weak_factory_(this) {
+  // This shouldn't be called twice.
+  DCHECK(!current_snapshot_);
+
+  // We only support 1.3+. There were library changes before this and we should
+  // use the new interface instead of the 1.2 one.
+  int randr_version_major = 0;
+  int randr_version_minor = 0;
+  if (XRRQueryVersion(xdisplay_, &randr_version_major, &randr_version_minor)) {
+    xrandr_version_ = randr_version_major * 100 + randr_version_minor;
+  }
+  // Need at least xrandr version 1.3.
+  if (xrandr_version_ >= 103) {
+    int error_base_ignored = 0;
+    XRRQueryExtension(xdisplay_, &xrandr_event_base_, &error_base_ignored);
+
+    if (ui::PlatformEventSource::GetInstance())
+      ui::PlatformEventSource::GetInstance()->AddPlatformEventDispatcher(this);
+    XRRSelectInput(xdisplay_, x_root_window_,
+                   RRScreenChangeNotifyMask | RROutputChangeNotifyMask |
+                       RRCrtcChangeNotifyMask);
+    SetDisplaysInternal(BuildDisplaysFromXRandRInfo());
+  } else {
+    SetDisplaysInternal(GetFallbackDisplayList());
+  }
+  SetDisplaySnapshot();
+}
+
+X11DisplayManager::~X11DisplayManager() {
+  if (xrandr_version_ >= 103 && ui::PlatformEventSource::GetInstance())
+    ui::PlatformEventSource::GetInstance()->RemovePlatformEventDispatcher(this);
+}
+
+void X11DisplayManager::SetObserver(Observer* observer) {
+  observer_ = observer;
+  if (current_mode_)
+    observer_->OnOutputReadyForUse();
+}
+
+void X11DisplayManager::GetDisplaysSnapshot(
+    display::GetDisplaysCallback callback) {
+  std::vector<display::DisplaySnapshot*> snapshot;
+  snapshot.push_back(current_snapshot_.get());
+  std::move(callback).Run(snapshot);
+}
+
+bool X11DisplayManager::CanDispatchEvent(const ui::PlatformEvent& event) {
+  // TODO(msisov, jkim): implement it
+  NOTIMPLEMENTED_LOG_ONCE();
+  return false;
+}
+
+uint32_t X11DisplayManager::DispatchEvent(const ui::PlatformEvent& event) {
+  // TODO(msisov, jkim): implement it
+  NOTIMPLEMENTED_LOG_ONCE();
+  return ui::POST_DISPATCH_NONE;
+}
+
+typedef XRRMonitorInfo* (*XRRGetMonitors)(::Display*, Window, bool, int*);
+typedef void (*XRRFreeMonitors)(XRRMonitorInfo*);
+
+PROTECTED_MEMORY_SECTION base::ProtectedMemory<XRRGetMonitors>
+    g_XRRGetMonitors_ptr;
+PROTECTED_MEMORY_SECTION base::ProtectedMemory<XRRFreeMonitors>
+    g_XRRFreeMonitors_ptr;
+
+std::vector<display::Display> X11DisplayManager::BuildDisplaysFromXRandRInfo() {
+  DCHECK(xrandr_version_ >= 103);
+  std::vector<display::Display> displays;
```

Comment:
why can't you create display snapshots here directly and avoid creating display::Display?